File: legacy/calculations.js
/* =========================================================================
 * calculations.js
 * Responsabilidade: todas as regras de cálculo de indicadores.
 * Este módulo é puro: recebe eventos e devolve números. Não conhece DOM,
 * LocalStorage nem gráficos — o que permite reaproveitá-lo no futuro em um
 * backend Node/Edge sem alteração.
 *
 * Fórmulas (TPM clássico):
 *   Tempo de Produção Planejado = Tempo Total do Turno − Paradas Planejadas
 *   Tempo Operacional           = TPP − Paradas Não Planejadas
 *   Disponibilidade             = Tempo Operacional ÷ TPP
 *   Performance                 = (Produção Total × Ciclo Ideal) ÷ Tempo Operacional
 *   Qualidade                   = Produção Aprovada ÷ Produção Total
 *   OEE                         = Disponibilidade × Performance × Qualidade
 *   Produção Aprovada           = Produção Total − Refugos
 * ========================================================================= */
(function (global) {
  'use strict';

  var OEE = (global.OEE = global.OEE || {});
  var Model = OEE.Model;

  /* ---------------------------------------------------------------------
   * Utilidades numéricas e de formatação
   * ------------------------------------------------------------------- */

  function seguro(n) { return (typeof n === 'number' && isFinite(n)) ? n : 0; }

  /* Divisão protegida: sem tempo/quantidade base, o indicador é indefinido
     (devolve null) em vez de 0 — o que evita "OEE 0%" falso em máquina sem
     tempo planejado. */
  function dividir(a, b) {
    if (!b || b <= 0) return null;
    return a / b;
  }

  /* Interseção entre o segmento [ini,fim] e a janela [janIni,janFim]. */
  function sobreposicaoMs(ini, fim, janIni, janFim) {
    var f = (fim === null || fim === undefined) ? janFim : fim;
    var a = Math.max(ini, janIni);
    var b = Math.min(f, janFim);
    return b > a ? b - a : 0;
  }

  function classeDoEstado(estadoId, config) {
    var override = config && config.classificacaoEstados && config.classificacaoEstados[estadoId];
    if (override) return override;
    var def = Model.ESTADOS[estadoId];
    return def ? def.classe : 'NAO_PLANEJADA';
  }

  function formatarPercentual(valor, casas) {
    if (valor === null || valor === undefined) return '—';
    var c = (casas === undefined) ? 1 : casas;
    return (valor * 100).toFixed(c).replace('.', ',') + '%';
  }

  function formatarNumero(valor, casas) {
    if (valor === null || valor === undefined) return '—';
    var c = (casas === undefined) ? 0 : casas;
    return Number(valor).toLocaleString('pt-BR', { minimumFractionDigits: c, maximumFractionDigits: c });
  }

  /* Duração legível: 1h 24min / 8min 30s */
  function formatarDuracao(segundos) {
    if (segundos === null || segundos === undefined || !isFinite(segundos)) return '—';
    var s = Math.max(0, Math.round(segundos));
    var h = Math.floor(s / 3600);
    var m = Math.floor((s % 3600) / 60);
    var r = s % 60;
    if (h > 0) return h + 'h ' + String(m).padStart(2, '0') + 'min';
    if (m > 0) return m + 'min ' + String(r).padStart(2, '0') + 's';
    return r + 's';
  }

  /* Cronômetro do estado atual: HH:MM:SS */
  function formatarCronometro(segundos) {
    var s = Math.max(0, Math.round(segundos || 0));
    var h = Math.floor(s / 3600);
    var m = Math.floor((s % 3600) / 60);
    var r = s % 60;
    return String(h).padStart(2, '0') + ':' + String(m).padStart(2, '0') + ':' + String(r).padStart(2, '0');
  }

  function formatarHora(ts) {
    if (!ts) return '—';
    var d = new Date(ts);
    return String(d.getHours()).padStart(2, '0') + ':' + String(d.getMinutes()).padStart(2, '0');
  }

  function formatarDataHora(ts) {
    if (!ts) return '—';
    var d = new Date(ts);
    return d.toLocaleDateString('pt-BR') + ' ' + formatarHora(ts);
  }

  /* ---------------------------------------------------------------------
   * Agregação de tempos por classe de estado
   * ------------------------------------------------------------------- */

  /**
   * Soma a duração dos eventos de estado dentro da janela informada.
   * Eventos abertos (fim === null) são contados até o fim da janela.
   *
   * @returns {Object} tempos em segundos, por estado e por classe.
   */
  function agregarTempos(eventosEstado, janIni, janFim, config) {
    var porEstado = {};
    var porClasse = { PRODUTIVO: 0, PLANEJADA: 0, NAO_PLANEJADA: 0 };
    var totalCoberto = 0;

    eventosEstado.forEach(function (ev) {
      var ms = sobreposicaoMs(ev.inicio, ev.fim, janIni, janFim);
      if (ms <= 0) return;
      var seg = ms / 1000;
      porEstado[ev.estado] = (porEstado[ev.estado] || 0) + seg;
      porClasse[classeDoEstado(ev.estado, config)] += seg;
      totalCoberto += seg;
    });

    return { porEstado: porEstado, porClasse: porClasse, totalCobertoSeg: totalCoberto };
  }

  /* ---------------------------------------------------------------------
   * Cálculo principal de OEE
   * ------------------------------------------------------------------- */

  /**
   * @param {Object} p
   *   p.eventosEstado   {Array}  eventos de estado da(s) máquina(s)
   *   p.eventosProducao {Array}  apontamentos de produção
   *   p.inicio, p.fim   {Number} janela de análise (ms)
   *   p.cicloIdealSeg   {Number} ciclo ideal médio (s/peça)
   *   p.config          {Object} configurações (classificação de estados)
   *
   * @returns {Object} indicadores + memória de cálculo + anomalias.
   */
  function calcularOEE(p) {
    var config = p.config || {};
    var janIni = p.inicio;
    var janFim = p.fim;
    var anomalias = [];

    var tempos = agregarTempos(p.eventosEstado || [], janIni, janFim, config);

    /* Tempo total considerado: apenas o período em que existe registro de
       estado para a máquina. Períodos sem registro não são atribuídos como
       perda — evita penalizar máquina recém-cadastrada. */
    var tempoTotalSeg = tempos.totalCobertoSeg;
    var paradaPlanejadaSeg = tempos.porClasse.PLANEJADA;
    var paradaNaoPlanejadaSeg = tempos.porClasse.NAO_PLANEJADA;

    /* TPP = tempo total − paradas planejadas */
    var tempoProducaoPlanejadoSeg = Math.max(0, tempoTotalSeg - paradaPlanejadaSeg);
    /* Tempo operacional = TPP − paradas não planejadas */
    var tempoOperacionalSeg = Math.max(0, tempoProducaoPlanejadoSeg - paradaNaoPlanejadaSeg);

    /* Produção no período */
    var producaoTotal = 0, refugo = 0, retrabalho = 0;
    (p.eventosProducao || []).forEach(function (ev) {
      if (ev.ts < janIni || ev.ts > janFim) return;
      producaoTotal += seguro(ev.qtdTotal);
      refugo += seguro(ev.qtdRefugo);
      retrabalho += seguro(ev.qtdRetrabalho);
    });
    var producaoAprovada = Math.max(0, producaoTotal - refugo);

    var cicloIdealSeg = p.cicloIdealSeg || 0;

    /* Indicadores */
    var disponibilidade = dividir(tempoOperacionalSeg, tempoProducaoPlanejadoSeg);
    var performance = dividir(producaoTotal * cicloIdealSeg, tempoOperacionalSeg);
    var qualidade = dividir(producaoAprovada, producaoTotal);

    /* Ciclo real médio observado. */
    var cicloRealSeg = dividir(tempoOperacionalSeg, producaoTotal);

    /* Validação: indicadores acima de 100% indicam inconsistência de
       apontamento (ciclo ideal errado, produção lançada em excesso ou
       parada não registrada). Registramos a anomalia e limitamos o valor
       usado no OEE, preservando o valor bruto para auditoria. */
    var performanceBruta = performance;
    if (performance !== null && performance > 1) {
      anomalias.push({
        indicador: 'Performance',
        valorBruto: performance,
        mensagem: 'Performance acima de 100%: revise o ciclo ideal, o apontamento de produção ou paradas não registradas.'
      });
      performance = 1;
    }
    var disponibilidadeBruta = disponibilidade;
    if (disponibilidade !== null && disponibilidade > 1) {
      anomalias.push({
        indicador: 'Disponibilidade',
        valorBruto: disponibilidade,
        mensagem: 'Disponibilidade acima de 100%: há sobreposição de eventos de estado no período.'
      });
      disponibilidade = 1;
    }
    if (qualidade !== null && qualidade > 1) qualidade = 1;

    var oee = (disponibilidade === null || performance === null || qualidade === null)
      ? null
      : disponibilidade * performance * qualidade;

    /* MTBF / MTTR considerando apenas paradas não planejadas. */
    var falhas = (p.eventosParada || []).filter(function (ev) {
      return !ev.planejada && sobreposicaoMs(ev.inicio, ev.fim, janIni, janFim) > 0;
    });
    var tempoFalhasSeg = falhas.reduce(function (acc, ev) {
      return acc + sobreposicaoMs(ev.inicio, ev.fim, janIni, janFim) / 1000;
    }, 0);
    var mtbfSeg = falhas.length ? tempoOperacionalSeg / falhas.length : null;
    var mttrSeg = falhas.length ? tempoFalhasSeg / falhas.length : null;

    return {
      janela: { inicio: janIni, fim: janFim },
      tempoTotalSeg: tempoTotalSeg,
      paradaPlanejadaSeg: paradaPlanejadaSeg,
      paradaNaoPlanejadaSeg: paradaNaoPlanejadaSeg,
      tempoProducaoPlanejadoSeg: tempoProducaoPlanejadoSeg,
      tempoOperacionalSeg: tempoOperacionalSeg,
      temposPorEstado: tempos.porEstado,
      producaoTotal: producaoTotal,
      producaoAprovada: producaoAprovada,
      refugo: refugo,
      retrabalho: retrabalho,
      cicloIdealSeg: cicloIdealSeg,
      cicloRealSeg: cicloRealSeg,
      disponibilidade: disponibilidade,
      performance: performance,
      qualidade: qualidade,
      oee: oee,
      disponibilidadeBruta: disponibilidadeBruta,
      performanceBruta: performanceBruta,
      qtdParadas: falhas.length,
      tempoFalhasSeg: tempoFalhasSeg,
      mtbfSeg: mtbfSeg,
      mttrSeg: mttrSeg,
      anomalias: anomalias
    };
  }

  /* ---------------------------------------------------------------------
   * Agrupamentos usados nos dashboards
   * ------------------------------------------------------------------- */

  /**
   * Calcula OEE para cada grupo (máquina, linha, turno, produto...).
   * @param chaveFn recebe a máquina e devolve a chave do grupo.
   */
  function oeePorGrupo(dataset, maquinas, janIni, janFim, chaveFn, rotuloFn) {
    var grupos = {};

    maquinas.forEach(function (maq) {
      var chave = chaveFn(maq);
      if (chave === null || chave === undefined) return;
      if (!grupos[chave]) {
        grupos[chave] = { chave: chave, rotulo: rotuloFn ? rotuloFn(chave, maq) : String(chave), maquinas: [] };
      }
      grupos[chave].maquinas.push(maq);
    });

    return Object.keys(grupos).map(function (k) {
      var g = grupos[k];
      var ids = g.maquinas.map(function (m) { return m.id; });
      var res = calcularOEE({
        eventosEstado: dataset.eventosEstado.filter(function (e) { return ids.indexOf(e.maquinaId) >= 0; }),
        eventosProducao: dataset.eventosProducao.filter(function (e) { return ids.indexOf(e.maquinaId) >= 0; }),
        eventosParada: dataset.eventosParada.filter(function (e) { return ids.indexOf(e.maquinaId) >= 0; }),
        inicio: janIni,
        fim: janFim,
        cicloIdealSeg: cicloIdealMedio(dataset, ids, janIni, janFim),
        config: dataset.config
      });
      return { chave: g.chave, rotulo: g.rotulo, maquinas: g.maquinas, indicadores: res };
    });
  }

  /**
   * Ciclo ideal médio ponderado pela quantidade produzida de cada produto.
   * Necessário quando o grupo mistura produtos com ciclos diferentes.
   */
  function cicloIdealMedio(dataset, maquinaIds, janIni, janFim) {
    var somaPeso = 0, somaCiclo = 0;
    dataset.eventosProducao.forEach(function (ev) {
      if (maquinaIds.indexOf(ev.maquinaId) < 0) return;
      if (ev.ts < janIni || ev.ts > janFim) return;
      var ordem = dataset.ordens.filter(function (o) { return o.id === ev.ordemId; })[0];
      var ciclo = ordem ? ordem.cicloIdealSeg : null;
      if (!ciclo) {
        var prod = dataset.produtos.filter(function (p) { return p.id === ev.produtoId; })[0];
        ciclo = prod ? prod.cicloIdealSeg : 0;
      }
      somaCiclo += ciclo * ev.qtdTotal;
      somaPeso += ev.qtdTotal;
    });
    if (somaPeso > 0) return somaCiclo / somaPeso;

    /* Sem produção no período: usa o ciclo ideal cadastrado nas máquinas. */
    var maquinas = dataset.maquinas.filter(function (m) { return maquinaIds.indexOf(m.id) >= 0; });
    if (!maquinas.length) return 0;
    return maquinas.reduce(function (a, m) { return a + (m.cicloIdealSeg || 0); }, 0) / maquinas.length;
  }

  /* Pareto de paradas: agrupa por motivo ou categoria e ordena por tempo. */
  function pareto(eventosParada, janIni, janFim, campo, rotuloFn) {
    var mapa = {};
    eventosParada.forEach(function (ev) {
      var ms = sobreposicaoMs(ev.inicio, ev.fim, janIni, janFim);
      if (ms <= 0) return;
      var chave = ev[campo] || 'Outros';
      if (!mapa[chave]) mapa[chave] = { chave: chave, rotulo: rotuloFn ? rotuloFn(chave) : String(chave), tempoSeg: 0, ocorrencias: 0 };
      mapa[chave].tempoSeg += ms / 1000;
      mapa[chave].ocorrencias += 1;
    });

    var lista = Object.keys(mapa).map(function (k) { return mapa[k]; })
      .sort(function (a, b) { return b.tempoSeg - a.tempoSeg; });

    var total = lista.reduce(function (a, i) { return a + i.tempoSeg; }, 0);
    var acumulado = 0;
    lista.forEach(function (i) {
      i.participacao = total ? i.tempoSeg / total : 0;
      acumulado += i.participacao;
      i.acumulado = acumulado;
    });
    return { itens: lista, totalSeg: total };
  }

  /* Série temporal de OEE em buckets (hora ou dia). */
  function serieOEE(dataset, maquinaIds, janIni, janFim, bucketMs) {
    var pontos = [];
    var estados = dataset.eventosEstado.filter(function (e) { return maquinaIds.indexOf(e.maquinaId) >= 0; });
    var producao = dataset.eventosProducao.filter(function (e) { return maquinaIds.indexOf(e.maquinaId) >= 0; });
    var paradas = dataset.eventosParada.filter(function (e) { return maquinaIds.indexOf(e.maquinaId) >= 0; });

    for (var t = janIni; t < janFim; t += bucketMs) {
      var fimBucket = Math.min(t + bucketMs, janFim);
      var res = calcularOEE({
        eventosEstado: estados, eventosProducao: producao, eventosParada: paradas,
        inicio: t, fim: fimBucket,
        cicloIdealSeg: cicloIdealMedio(dataset, maquinaIds, t, fimBucket),
        config: dataset.config
      });
      pontos.push({ inicio: t, fim: fimBucket, indicadores: res });
    }
    return pontos;
  }

  /* Produção agregada por bucket de tempo (usada em "produção por hora"). */
  function producaoPorBucket(eventosProducao, janIni, janFim, bucketMs) {
    var buckets = [];
    for (var t = janIni; t < janFim; t += bucketMs) {
      buckets.push({ inicio: t, fim: Math.min(t + bucketMs, janFim), total: 0, aprovada: 0, refugo: 0 });
    }
    eventosProducao.forEach(function (ev) {
      if (ev.ts < janIni || ev.ts > janFim) return;
      var idx = Math.floor((ev.ts - janIni) / bucketMs);
      if (idx < 0 || idx >= buckets.length) return;
      buckets[idx].total += seguro(ev.qtdTotal);
      buckets[idx].refugo += seguro(ev.qtdRefugo);
      buckets[idx].aprovada += seguro(ev.qtdTotal) - seguro(ev.qtdRefugo);
    });
    return buckets;
  }

  /* Classificação simples para semáforo de OEE. */
  function faixaOEE(valor, meta) {
    if (valor === null || valor === undefined) return 'indefinido';
    var m = meta || 0.75;
    if (valor >= m) return 'bom';
    if (valor >= m * 0.8) return 'atencao';
    return 'critico';
  }

  OEE.Calc = {
    calcularOEE: calcularOEE,
    agregarTempos: agregarTempos,
    oeePorGrupo: oeePorGrupo,
    cicloIdealMedio: cicloIdealMedio,
    pareto: pareto,
    serieOEE: serieOEE,
    producaoPorBucket: producaoPorBucket,
    sobreposicaoMs: sobreposicaoMs,
    classeDoEstado: classeDoEstado,
    faixaOEE: faixaOEE,
    dividir: dividir,
    fmt: {
      percentual: formatarPercentual,
      numero: formatarNumero,
      duracao: formatarDuracao,
      cronometro: formatarCronometro,
      hora: formatarHora,
      dataHora: formatarDataHora
    }
  };
})(window);

File: apps/api/src/oee/domain/calculations.py
"""Fórmulas TPM de OEE — portadas de legacy/calculations.js (funções puras)."""

from __future__ import annotations

from typing import Any

from oee.domain.catalog import classe_do_estado


def seguro(n: Any) -> float:
    try:
        v = float(n)
        return v if v == v and v not in (float("inf"), float("-inf")) else 0.0
    except (TypeError, ValueError):
        return 0.0


def dividir(a: float | None, b: float | None) -> float | None:
    if not b or b <= 0:
        return None
    return (a or 0) / b


def sobreposicao_ms(ini: int | None, fim: int | None, jan_ini: int, jan_fim: int) -> int:
    f = jan_fim if fim is None else fim
    a = max(ini or 0, jan_ini)
    b = min(f, jan_fim)
    return b - a if b > a else 0


def agregar_tempos(
    eventos_estado: list[dict[str, Any]],
    jan_ini: int,
    jan_fim: int,
    config: dict[str, Any] | None = None,
) -> dict[str, Any]:
    por_estado: dict[str, float] = {}
    por_classe = {"PRODUTIVO": 0.0, "PLANEJADA": 0.0, "NAO_PLANEJADA": 0.0}
    total = 0.0
    for ev in eventos_estado:
        ms = sobreposicao_ms(ev.get("inicio"), ev.get("fim"), jan_ini, jan_fim)
        if ms <= 0:
            continue
        seg = ms / 1000.0
        estado = ev.get("estado")
        por_estado[estado] = por_estado.get(estado, 0.0) + seg
        por_classe[classe_do_estado(estado, config)] += seg
        total += seg
    return {"por_estado": por_estado, "por_classe": por_classe, "total_coberto_seg": total}


def calcular_oee(p: dict[str, Any]) -> dict[str, Any]:
    config = p.get("config") or {}
    jan_ini = p["inicio"]
    jan_fim = p["fim"]
    anomalias: list[dict[str, Any]] = []

    tempos = agregar_tempos(p.get("eventos_estado") or [], jan_ini, jan_fim, config)
    tempo_total = tempos["total_coberto_seg"]
    parada_planejada = tempos["por_classe"]["PLANEJADA"]
    parada_nao_planejada = tempos["por_classe"]["NAO_PLANEJADA"]
    tpp = max(0.0, tempo_total - parada_planejada)
    operacional = max(0.0, tpp - parada_nao_planejada)

    producao_total = 0.0
    refugo = 0.0
    retrabalho = 0.0
    for ev in p.get("eventos_producao") or []:
        ts = ev.get("ts")
        if ts is None or ts < jan_ini or ts > jan_fim:
            continue
        producao_total += seguro(ev.get("qtd_total", ev.get("qtdTotal")))
        refugo += seguro(ev.get("qtd_refugo", ev.get("qtdRefugo")))
        retrabalho += seguro(ev.get("qtd_retrabalho", ev.get("qtdRetrabalho")))
    aprovada = max(0.0, producao_total - refugo)

    ciclo_ideal = p.get("ciclo_ideal_seg") or 0.0
    disponibilidade = dividir(operacional, tpp)
    performance = dividir(producao_total * ciclo_ideal, operacional)
    qualidade = dividir(aprovada, producao_total)
    ciclo_real = dividir(operacional, producao_total)

    performance_bruta = performance
    if performance is not None and performance > 1:
        anomalias.append(
            {
                "indicador": "Performance",
                "valor_bruto": performance,
                "mensagem": "Performance acima de 100%: revise o ciclo ideal, o apontamento de produção ou paradas não registradas.",
            }
        )
        performance = 1.0
    disponibilidade_bruta = disponibilidade
    if disponibilidade is not None and disponibilidade > 1:
        anomalias.append(
            {
                "indicador": "Disponibilidade",
                "valor_bruto": disponibilidade,
                "mensagem": "Disponibilidade acima de 100%: há sobreposição de eventos de estado no período.",
            }
        )
        disponibilidade = 1.0
    if qualidade is not None and qualidade > 1:
        qualidade = 1.0

    oee = None
    if disponibilidade is not None and performance is not None and qualidade is not None:
        oee = disponibilidade * performance * qualidade

    falhas = [
        ev
        for ev in (p.get("eventos_parada") or [])
        if not ev.get("planejada") and sobreposicao_ms(ev.get("inicio"), ev.get("fim"), jan_ini, jan_fim) > 0
    ]
    tempo_falhas = sum(sobreposicao_ms(ev.get("inicio"), ev.get("fim"), jan_ini, jan_fim) / 1000.0 for ev in falhas)
    mtbf = dividir(operacional, float(len(falhas))) if falhas else None
    mttr = dividir(tempo_falhas, float(len(falhas))) if falhas else None

    return {
        "janela": {"inicio": jan_ini, "fim": jan_fim},
        "tempo_total_seg": tempo_total,
        "parada_planejada_seg": parada_planejada,
        "parada_nao_planejada_seg": parada_nao_planejada,
        "tempo_producao_planejado_seg": tpp,
        "tempo_operacional_seg": operacional,
        "tempos_por_estado": tempos["por_estado"],
        "producao_total": producao_total,
        "producao_aprovada": aprovada,
        "refugo": refugo,
        "retrabalho": retrabalho,
        "ciclo_ideal_seg": ciclo_ideal,
        "ciclo_real_seg": ciclo_real,
        "disponibilidade": disponibilidade,
        "performance": performance,
        "qualidade": qualidade,
        "oee": oee,
        "disponibilidade_bruta": disponibilidade_bruta,
        "performance_bruta": performance_bruta,
        "qtd_paradas": len(falhas),
        "tempo_falhas_seg": tempo_falhas,
        "mtbf_seg": mtbf,
        "mttr_seg": mttr,
        "anomalias": anomalias,
    }


def ciclo_ideal_medio(
    dataset: dict[str, Any],
    maquina_ids: list[str],
    jan_ini: int,
    jan_fim: int,
) -> float:
    ids = set(maquina_ids)
    soma_peso = 0.0
    soma_ciclo = 0.0
    ordens = {o["id"]: o for o in dataset.get("ordens") or []}
    produtos = {p["id"]: p for p in dataset.get("produtos") or []}
    for ev in dataset.get("eventos_producao") or []:
        if ev.get("maquina_id") not in ids:
            continue
        ts = ev.get("ts")
        if ts is None or ts < jan_ini or ts > jan_fim:
            continue
        ordem = ordens.get(ev.get("ordem_id"))
        ciclo = ordem.get("ciclo_ideal_seg") if ordem else None
        if not ciclo:
            prod = produtos.get(ev.get("produto_id"))
            ciclo = prod.get("ciclo_ideal_seg") if prod else 0
        qtd = seguro(ev.get("qtd_total", ev.get("qtdTotal")))
        soma_ciclo += (ciclo or 0) * qtd
        soma_peso += qtd
    if soma_peso > 0:
        return soma_ciclo / soma_peso
    maquinas = [m for m in dataset.get("maquinas") or [] if m["id"] in ids]
    if not maquinas:
        return 0.0
    return sum(m.get("ciclo_ideal_seg") or 0 for m in maquinas) / len(maquinas)


def oee_por_grupo(
    dataset: dict[str, Any],
    maquinas: list[dict[str, Any]],
    jan_ini: int,
    jan_fim: int,
    chave_fn,
    rotulo_fn=None,
) -> list[dict[str, Any]]:
    grupos: dict[str, dict[str, Any]] = {}
    for maq in maquinas:
        chave = chave_fn(maq)
        if chave is None:
            continue
        if chave not in grupos:
            grupos[chave] = {
                "chave": chave,
                "rotulo": rotulo_fn(chave, maq) if rotulo_fn else str(chave),
                "maquinas": [],
            }
        grupos[chave]["maquinas"].append(maq)

    out = []
    for g in grupos.values():
        ids = [m["id"] for m in g["maquinas"]]
        idset = set(ids)
        res = calcular_oee(
            {
                "eventos_estado": [e for e in dataset.get("eventos_estado") or [] if e.get("maquina_id") in idset],
                "eventos_producao": [e for e in dataset.get("eventos_producao") or [] if e.get("maquina_id") in idset],
                "eventos_parada": [e for e in dataset.get("eventos_parada") or [] if e.get("maquina_id") in idset],
                "inicio": jan_ini,
                "fim": jan_fim,
                "ciclo_ideal_seg": ciclo_ideal_medio(dataset, ids, jan_ini, jan_fim),
                "config": dataset.get("config") or {},
            }
        )
        out.append({**g, "indicadores": res})
    return out


def pareto(
    eventos_parada: list[dict[str, Any]],
    jan_ini: int,
    jan_fim: int,
    campo: str,
    rotulo_fn=None,
) -> dict[str, Any]:
    mapa: dict[str, dict[str, Any]] = {}
    for ev in eventos_parada:
        ms = sobreposicao_ms(ev.get("inicio"), ev.get("fim"), jan_ini, jan_fim)
        if ms <= 0:
            continue
        chave = ev.get(campo) or "Outros"
        if chave not in mapa:
            mapa[chave] = {
                "chave": chave,
                "rotulo": rotulo_fn(chave) if rotulo_fn else str(chave),
                "tempo_seg": 0.0,
                "ocorrencias": 0,
            }
        mapa[chave]["tempo_seg"] += ms / 1000.0
        mapa[chave]["ocorrencias"] += 1
    lista = sorted(mapa.values(), key=lambda i: i["tempo_seg"], reverse=True)
    total = sum(i["tempo_seg"] for i in lista)
    acumulado = 0.0
    for i in lista:
        i["participacao"] = (i["tempo_seg"] / total) if total else 0.0
        acumulado += i["participacao"]
        i["acumulado"] = acumulado
    return {"itens": lista, "total_seg": total}


def serie_oee(
    dataset: dict[str, Any],
    maquina_ids: list[str],
    jan_ini: int,
    jan_fim: int,
    bucket_ms: int,
) -> list[dict[str, Any]]:
    idset = set(maquina_ids)
    estados = [e for e in dataset.get("eventos_estado") or [] if e.get("maquina_id") in idset]
    producao = [e for e in dataset.get("eventos_producao") or [] if e.get("maquina_id") in idset]
    paradas = [e for e in dataset.get("eventos_parada") or [] if e.get("maquina_id") in idset]
    pontos = []
    t = jan_ini
    while t < jan_fim:
        fim_b = min(t + bucket_ms, jan_fim)
        res = calcular_oee(
            {
                "eventos_estado": estados,
                "eventos_producao": producao,
                "eventos_parada": paradas,
                "inicio": t,
                "fim": fim_b,
                "ciclo_ideal_seg": ciclo_ideal_medio(dataset, maquina_ids, t, fim_b),
                "config": dataset.get("config") or {},
            }
        )
        pontos.append({"inicio": t, "fim": fim_b, "indicadores": res})
        t += bucket_ms
    return pontos


def producao_por_bucket(
    eventos_producao: list[dict[str, Any]],
    jan_ini: int,
    jan_fim: int,
    bucket_ms: int,
) -> list[dict[str, Any]]:
    buckets = []
    t = jan_ini
    while t < jan_fim:
        buckets.append(
            {
                "inicio": t,
                "fim": min(t + bucket_ms, jan_fim),
                "total": 0.0,
                "aprovada": 0.0,
                "refugo": 0.0,
            }
        )
        t += bucket_ms
    for ev in eventos_producao:
        ts = ev.get("ts")
        if ts is None or ts < jan_ini or ts > jan_fim:
            continue
        idx = int((ts - jan_ini) // bucket_ms)
        if idx < 0 or idx >= len(buckets):
            continue
        qtd = seguro(ev.get("qtd_total", ev.get("qtdTotal")))
        ref = seguro(ev.get("qtd_refugo", ev.get("qtdRefugo")))
        buckets[idx]["total"] += qtd
        buckets[idx]["refugo"] += ref
        buckets[idx]["aprovada"] += qtd - ref
    return buckets


def faixa_oee(valor: float | None, meta: float = 0.75) -> str:
    if valor is None:
        return "indefinido"
    if valor >= meta:
        return "bom"
    if valor >= meta * 0.8:
        return "atencao"
    return "critico"
